test/lib/jdk/test/lib/format/ArrayDiff.java

```diff
@@ -0,0 +1,210 @@
+ * printed, element in String[] arrays are additionally surrounded with quotes.
+ * Additional formatting parameters, like maximum allowed width and number of
+ * elements printed before difference, can be specified.
+ *
+ * <p>Output examples:
+ *
+ * <p> two int arrays: </p>
+ * <pre>
+ * Arrays differ starting from [index: 4]:
+ *         ... 3, 4,   5, 6, 7]
+ *         ... 3, 4, 225, 6, 7]
+ *                  ^^^^
+ * </pre>
+ * <p> two String arrays: </p>
+ * <pre>
+ * Arrays differ starting from [index: 2]:
+ *         ["first", "second",     "third", "u\nprintable"]
+ *         ["first", "second", "incorrect", "u\nprintable"]
+ *                            ^^^^^^^^^^^^
+ * </pre>
+ * <p> two char arrays arrays: </p>
+ * <pre>
+ * Arrays differ starting from [index: 7]:
+ *         ... &#92;u0001, &#92;u0002, &#92;u0007, a, b, \n, ...
+ *         ... &#92;u0001, &#92;u0002,      }, a, b, \n, ...
+ *                            ^^^^^^^
+ * </pre>
+ */
+public class ArrayDiff<E> implements Diff {
+
+    private int failureIdx;
+    private final int maxWidth;
+    private final int contextBefore;
+
+    private final ArrayCodec<E> first;
+    private final ArrayCodec<E> second;
+
+    private ArrayDiff(ArrayCodec<E> first, ArrayCodec<E> second,
+                      int width, int getContextBefore) {
+        this.first = first;
+        this.second = second;
+        this.maxWidth = width;
+        this.contextBefore = getContextBefore;
+        failureIdx = first.findMismatchIndex(second);
+    }
+
+    /**
+     * Creates an ArrayDiff fom two arrays and default limits. The given arguments must be of the same
+     * component type.
+     *
+     * @param first the first array
+     * @param second the second array
+     * @return an ArrayDiff instance for the two arrays
+     */
+    public static ArrayDiff of(Object first, Object second) {
+        return ArrayDiff.of(first, second, Diff.Defaults.WIDTH, Diff.Defaults.CONTEXT_BEFORE);
+    }
+
+    /**
+     * Creates an ArrayDiff fom two arrays with the given limits. The given arguments must be of the same
+     * component type.
+     *
+     * @param first the first array
+     * @param second the second array
+     * @param width the maximum allowed width in characters for the formatting
+     * @param contextBefore maximum number of elements to print before those that differ
+     * @throws IllegalArgumentException if component types of arrays is not supported or are not the same
+     * @return an ArrayDiff instance for the two arrays and formatting parameters provided
+     */
+    public static ArrayDiff of(Object first, Object second, int width, int contextBefore) {
+        if (first == null) {
+            throw new IllegalArgumentException("First array argument for ArrayDiff is null");
+        }
+
+        if (second == null) {
+            throw new IllegalArgumentException("Second array argument for ArrayDiff is null");
+        }
```

Comment:
it's more common (and less surprising) to throw NPE in such cases, preferably by using `Objects::requireNonNull`

Reply:
Haven't known about the `requireNonNull`, thanks. I was thinking about the NPE myself, just wanted to show up which array was null. Fixed in the ba6e5411b4de5cf8b9230b7cdb7a518900eadc4b.